### ISSUE TITLE
Improve markers clustering experience

### ISSUE THREAD

**Issue by livthomas** [mapbox/mapbox-gl-js#8478]
## Motivation

I am trying to add clustering of HTML markers to my application. And since Mapbox does not directly support it, we need to implement it on our own. [One of the solutions](https://github.com/mapbox/mapbox-gl-js/issues/4491#issuecomment-501442036) is using built-in clustering but not showing individual "markers" (just cluster circles) and re-rendering HTML markers on every `moveend` event.

This works, however, sometimes there is a situation when there is a longer transition (e.g. after clicking on a cluster circle) and the cluster is not recalculated before `moveend` event is emitted. To work around this, I am calling the following line in `moveend` event callback:

```typescript
this.mapboxMap.once('idle', () => this.redrawMarkers());
```

But the `idle` event is emitted only after all transition animations have completed and all tiles have loaded. The problem is it usually takes a while to load the tiles and so marker clusters are not re-rendered immediately after the transition ended but after another 1-2 seconds.

## Design Alternatives

I can see many possible solutions of this problem:

1. [Add support for HTML markers clustering.](https://github.com/mapbox/mapbox-gl-js/issues/4491)
2. Add support for drawing arbitrary marker shapes (not only dots or circles) using WebGL.
2. Emit `moveend` event only after clusters are re-calculated.
3. Add some new event which will be emitted after clusters are re-calculated.
4. [Add some new event which will be emitted after transition animation have completed.](https://github.com/mapbox/mapbox-gl-js/issues/1482) (But I am not sure if this one will help.)

## Design

I think the best solution of this problem would be to implement one of the first two options. But I can see that it will not be that easy so one of 3-5 is probably a better solution for now.

**Comment by mourner** [comment]
Are you seeing this problem in the [official HTML marker example](https://docs.mapbox.com/mapbox-gl-js/example/cluster-html/)? If so, can you describe the exact steps to trigger the issue?

**Comment by livthomas** [comment]
@mourner No, it is in my custom implementation which is based on some of the ideas from [HTML clustering issue comments](https://github.com/mapbox/mapbox-gl-js/issues/4491). In fact, it is exactly the opposite of what you linked here. In my application, the markers themselves are HTML elements and cluster circles are simply a circle layer.

I am trying to put simplified code to JS Bin. But our application is quite complex (written in Angular) and it is quite hard to extract all the code from it to vanilla JavaScript while keeping all the functionality responsible for rendering and clustering of markers.

**Comment by mourner** [comment]
Can you set up a minimal test case (e.g. on JSFiddle) to demonstrate the issue here? No need to extract the whole code, the simplest example that can show the problem you're having would suffice.

**Comment by livthomas** [comment]
@mourner I have finally managed to put together a reproducer. It uses some data structures from our application so I hope it is not too complicated.

https://jsbin.com/yukumuyoha/edit?html,output

Try clicking on a marker cluster and see that the markers will not show up immediately but only after you move the map.

Then try to uncomment the following line in the code (it is there twice):
```javascript
map.once('idle', () => redrawMarkers());
```

After that the markers will always show up after clicking on a marker cluster, however, they will show up only after all map tiles have been loaded.

I would need a callback that would be called immediately after marker clusters are recalculated so I do not need to wait for all tiles to be loaded. That would be much better user experience.

**Comment by livthomas** [comment]
@mourner Sorry, I have shared the old JS Bin. You can find the new one here:

https://jsbin.com/vigoweduro/edit?html,output

**Comment by luchux** [comment]
I am facing the same issue that @livthomas 

@mourner  basically this happens when you need to use builtin functionality of clustering with custom html markers. You end up removing and adding markers from the map, based in their (in-cluster status).

**Comment by cetooley** [comment]
I am encountering this error using a similar marker clustering setup as here: https://blog.mapbox.com/clustering-properties-with-mapbox-and-html-markers-bb353c8662ba

It seems to me like the "moveend" event happens before any clustering or unclustering of features. If there was a set of "cluster" events which fired whenever there was a change to clusters I think this could be solved easily.

**Comment by cetooley** [comment]
I can get the issue to show on the [official documentation](https://docs.mapbox.com/mapbox-gl-js/example/cluster-html/) sometimes too, using Firefox 73.0. It's possibly related to the number of clusters on the screen - our implementation has a large number of clusters.

To attempt to reproduce:
Zoom using the mouse scroller by small increments until you get something like this:
![image](https://user-images.githubusercontent.com/33706076/74469884-534c1000-4e52-11ea-88b5-99971a197690.png)

Drag the map a bit and some clusters will disappear to give you this:
![image](https://user-images.githubusercontent.com/33706076/74469927-6828a380-4e52-11ea-9bc6-e93db8cc519a.png)

**Comment by arenarium-dev** [comment]
Hey, dont know if its still relevant, but Ive been building a plugin for mapbox and other map providers that enables you to organize a large number of custom markers.

![Image](https://github.com/user-attachments/assets/fd99d870-de50-4759-ba80-6de97923fb62)

You can chek it out at: [https://arenarium.dev/?demo=bookings](https://arenarium.dev/?demo=bookings)
